usr/src/uts/common/io/overlay/overlay_router.c

```diff
@@ -0,0 +1,2389 @@
+
+	switch (sa->sa_family) {
+	case AF_INET:
+		if (sin->sin_addr.s_addr == INADDR_ANY)
+			return (B_TRUE);
+		break;
+	case AF_INET6:
+		if (IN6_IS_ADDR_UNSPECIFIED(&sin6->sin6_addr))
+			return (B_TRUE);
+		break;
+	default:
+		cmn_err(CE_PANIC, "unexpected address family value");
+	}
+
+	return (B_FALSE);
+}
+
+int
+overlay_route(overlay_dev_t *odd, overlay_net_t *ont, overlay_pkt_t *pkt,
+    struct sockaddr *addr, socklen_t *lenp)
+{
+	overlay_router_t *orr = odd->odd_router;
+	overlay_routetab_t *rtab = NULL;
+	overlay_target_t *ott = NULL;
+	int ret = 0;
+
+	/*
+	 * If we're routing, the destination MAC and vlan of the packet
+	 * should agree, otherwise we drop the packet.
+	 */
+	if (OPKT_VLAN(pkt) != ont->ont_vlan)
+		return (OVERLAY_TARGET_DROP);
+
+	switch (pkt->op_mhi.mhi_bindsap) {
+	case ETHERTYPE_IP:
+	case ETHERTYPE_IPV6:
+		break;
+	default:
+		/*
+		 * Drop any non IPv4 or IPv6 packets directed to the router
+		 * MAC.
+		 */
+		return (OVERLAY_TARGET_DROP);
+	}
+
+	ott = odd->odd_target;
+	if (ott == NULL)
+		return (OVERLAY_TARGET_DROP);
+
+	/*
+	 * No routing table, and the packet was sent to the router MAC,
+	 * we drop.
+	 */
+	rtab = overlay_get_rtab(orr, ont);
+	if (rtab == NULL)
+		return (OVERLAY_TARGET_DROP);
+
+	/*
+	 * Find the target address+port for this packet in the routing table.
+	 * If no target was found, we drop the packet.
+	 */
+	if (!overlay_router_get_target(rtab, pkt, addr, lenp)) {
+		ret = OVERLAY_TARGET_DROP;
+		goto done;
+	}
+
+	/*
+	 * If the target is the 'local' destination (the target is the
+	 * any/unspecified address, we do a VL3->VL2 lookup and route within
+	 * the vnet.
+	 */
+	if (is_target_local(addr)) {
+		ret = overlay_target_lookup(odd, pkt, B_TRUE, addr, lenp);
+	}
+
+	/* XXX: Any additional handling (e.g. NAT, ...) would go here */
+
+done:
+	overlay_route_tbl_rele(rtab);
+	return (ret);
+}
+
+static int
+overlay_cmp_net_vlan(const void *a, const void *b)
+{
+	const overlay_net_t *l = a;
+	const overlay_net_t *r = b;
+
+	if (l->ont_vlan < r->ont_vlan)
+		return (-1);
+	if (l->ont_vlan > r->ont_vlan)
+		return (1);
+	return (0);
+}
+
+static int
+overlay_cmp_net_mac(const void *a, const void *b)
+{
+	const overlay_net_t *l = a;
+	const overlay_net_t *r = b;
+
+	for (uint_t i = 0; i < ETHERADDRL; i++) {
+		if (l->ont_mac[i] > r->ont_mac[i])
+			return (1);
+		if (l->ont_mac[i] < r->ont_mac[i])
+			return (-1);
+	}
+
+	return (0);
+}
+
+static int
+overlay_cmp_net_v4(const void *a, const void *b)
+{
+	const overlay_net_t *l_net = a;
+	const overlay_net_t *r_net = b;
+	const uint32_t l = ntohl((uint32_t)l_net->ont_net);
+	const uint32_t r = ntohl((uint32_t)r_net->ont_net);
+
+	if (l < r)
+		return (-1);
+	if (l > r)
+		return (1);
+	return (0);
+}
+
+static int
+overlay_cmp_net_v6(const void *a, const void *b)
+{
+	const overlay_net_t *l = a;
+	const overlay_net_t *r = b;
+	const struct in6_addr *laddr = &l->ont_netv6;
+	const struct in6_addr *raddr = &r->ont_netv6;
+
```

Comment:
It's doing a -erpbyte (instead of per 32-bit integer) comparison, so it shouldn't be necessary. If we want to compare 32 or 64 bits at a time, then I believe it would.